Run terraform fmt as githook and GH Action

ronaldheft-usds opened the pull request:
This PR brings some sanity to our Terraform formatting by using [terraform fmt](https://www.terraform.io/docs/cli/commands/fmt.html).

## Changes
- Adds a pre-commit githook that validates all Terraform files are formatted per `terraform fmt` specs before committing
  - The pre-commit githook provides details for fixing the files, including running the hook with the `--fix` option
- Runs a verification of Terraform formatting as part of GitHub Action as an added failsafe
- Formats all existing Terraform files
  - This will make for some merge conflicts, but it's a necessary evil
- Updates our `.editorconfig` to help the IDE make the same formatting choice at `terraform fmt`

## Checklist

### Testing
- [x] Tested locally?
- [ ] Ran `quickTest all`?
- [ ] Ran `./prime test` against local Docker ReportStream container?
- [ ] Downloaded a file from `http://localhost:7071/api/download`?
- [ ] Added tests?

### Specific Security-related subjects a reviewer should pay specific attention to

- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [x] Includes a summary of what a code reviewer should verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Fixes
*List GitHub issues this PR fixes*
- #issue

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

td-usds commented on .environment/terraform-fmt/run-terraform-fmt.sh:
```diff
@@ -0,0 +1,47 @@
+
+if [[ ${RC?} != 0 ]]; then
+    error "(return code=${RC?}) Your Terraform files are not formatted."
+    error "\`terraform fmt\` has been run and the changes can be committed to comply with formatting requirements."
```
Any changes performed by `terraform fmt` are not going to be staged and so must be added to the user's staged changes:

```suggestion
    error "\`terraform fmt\` has been run and the changes can be committed to comply with formatting requirements."
    error "Please validate the output from terraform fmt and then git add it to your staged changes."
```

There's a subtle bug here: if I commit non-compliant code that triggers this, it will format my code thus making it compliant. But, the formatted - and thus compliant - code is not yet staged. However, when I try to recommit - without doing _any_thing else, the hook will now pass because it checks the content of the files listed in your staging as it is in your working directory.

So I think it may make more sense to NOT try to format for the user and tell them to do the formatting (and then any subsequently required staging/git adding) themselves.

My suggestion would thus actually be to NOT try to format for the user!

ronaldheft-usds replied:
Agreed. Updated the hook to not automatically format.

td-usds commented on .environment/terraform-fmt/run-terraform-fmt.sh:
```diff
@@ -0,0 +1,47 @@
+terraform_check_fmt
+RC=$?
+
+terraform_fmt
```
Would it make more sense to condition execution of `terraform_fmt` on the value of `${RC?}` (and then _not_ do the check inside `terraform_fmt`, thus making it callable from anywhere without having to set the `${TERRAFORM_NEEDS_FMT}=1`)  or is there another reason why `terraform_fmt` is the one doing the check?

```suggestion
if [[ ${RC?} != 0 ]]; then
    terraform_fmt
fi
```

ronaldheft-usds replied:
Refactored the code so this is no longer an issue.

td-usds commented on .environment/terraform-fmt/run-terraform-fmt.sh:
```diff
@@ -0,0 +1,47 @@
+}
+
+function terraform_fmt() {
+    if [[ ${TERRAFORM_NEEDS_FMT?} != 0 ]]; then
```
See comment above to move the check outside of this function

ronaldheft-usds replied:
Refactored the code so this is no longer an issue.

td-usds commented on operations/app/src/modules/front_door/variables.tf:
```diff
@@ -1,29 +1,29 @@
 
 variable "https_cert_names" {
-  type = list
+  type        = list(any)
```
Is this supposed to be `any`? Should it be `object` or `string` instead?

ronaldheft-usds replied:
Fixed

td-usds commented on operations/app/src/modules/sftp_container/main.tf:
```diff
@@ -53,7 +53,7 @@ resource "azurerm_container_group" "sftp_container" {
     ignore_changes = [
       network_profile_id,
-      container[0].volume[0]]
+    container[0].volume[0]]
```
This is a weird formatting it did :/

td-usds commented on operations/app/src/modules/key_vault/main.tf:
```diff
@@ -1,234 +1,234 @@
+  object_id    = local.frontdoor_object_id
+
+  secret_permissions = [
+  "Get"]
```
What does the formatting do if you make this:
```hcl
secret_permissions = [
  "Get"
]
```
??

That might be a better way to do lists.

ronaldheft-usds replied:
Agreed, I fixed up all of these instances throughout the formatted code and ensured the last item has a trailing `,`.

td-usds reviewed (changes requested):
There's a subtle bug here: if I try to commit non-compliant code that triggers this, it will format my code for me ('magic') thus making it compliant and then block the commit from going in.

This is cool, but...

The formatted - and thus compliant - code is not yet staged. However, when I try to recommit without doing _any_thing else, the hook will now pass because it checks the content of the files listed in your staging as it is in your working directory. This is because the files checked are now compliant, but that's not what'll be committed. What will be committed is what is staged. It's the same 'set of file names' but not 'the same content'!

So I think it may make more sense to NOT try to format for the user and tell them to do the formatting (and then any subsequently required staging/git adding) themselves.

My suggestion would thus actually be to NOT try to format for the user!

ronaldheft-usds commented on .editorconfig:
```diff
@@ -6,5 +6,5 @@ indent_style = space
 tab_width = 4
 
-[{*.yaml,*.yml}]
+[{*.yaml,*.yml, *.tf}]
```
This matches what `terraform fmt` does.

ronaldheft-usds commented:
@td-usds Ready for review again. I took your suggestions, cleaned up the auto-formatting, and added a `.editorconfig` adjustment to help IDEs match the formatting.
